dra-evolution/README.md

```diff
@@ -162,21 +186,26 @@ claim types.
 
 Claim and allocation types are found in [claim_types.go](pkg/api/claim_types.go);
 individual types and fields are described in detail there in the comments.
+Capacity types are in [capacity_types.go](pkg/api/capacity_types.go). A quota
+mechanism is defined in [quota_types.go](pkg/api/quota_types.go).
+
+Vendors and administrators create `ResourceClass` resources to pre-configure
+various options for claims. Depending on what gets set in a class, users can:
+- Ask for exactly the set of devices pre-defined in a class.
+- Add additional configuration to their claim. This configuration is
+  passed down to the driver as coming from an admin, so it may control
+  options that normal users must not set themselves.
+- Restrict the choice of devices via additional constraints.
 
-Vendors and administrators create `DeviceClass` resources to pre-configure
-various options for claims. DeviceClass resources come in two varieties:
-- Ordinary or "leaf" classes that represent devices managed by a specific
-  driver, along with some optional selection constraints and configuration.
-- "Meta" or "Group" or "Aggregate" or "Composition" classes that use a label
-  selector to identify a *set* of leaf classes. This allows a claim to be
-  satistfied by one of many classes.
+Classes are not necessarily associated with a single vendor. Whether they are
```

Comment:
There is a third reason for classes:
- set configuration options that non-privileged users are not allowed to set

I don't know whether that is a strong enough reason to have them. But if we add a `resourceClassName` field later, it would be ignored by an older scheduler, so adding it later is going to be harder.

> equivalence (class of classes, which has never come up for volumes, but a cluster of hetergeneous nodes SEEMS plausible)

I'm not sure I understand what is meant with this. Is this a weaker form of a vendor-independent "give me a GPU" class?

> giving admins a way to express the bounding set of devices (IFF all claims must use classes)

My thinking was that this could be enforced with mutating webhooks, i.e. no need to have it in core Kubernetes. I might be wrong and it really has to be a mandatory field because a webhook would have to interpret the request to determine which device is meant, which might not be possible.

I'm not a fan of making classes mandatory. If we did this and then had "restricted" classes and "less restricted classes" in the cluster, we need access control for the classes. IMHO a better solution is to implement a quota mechanism that can be used to prevent the usage of certain devices per namespace, instead of doing it indirectly through classes.

Reply:
> Thus, claims should require a class, 

To me what you said sounds like "should be allowed to use a class". If an admin decides to provide those classes, fine, but if there is no class that does what a user needs, why force them to reference one?

> Furthermore, we should have a way for the admin to control which users (or namespaces) may consume which classes

What is in a class that a user cannot do themselves? I mentioned "privileged configuration" and @thockin correctly pointed out that we would need a specific example for that. I don't have any at hand, so perhaps this really isn't important (yet).

> Classes do also provide a major UX convenience

That remains to be seen. At this point we don't have evidence that this really will be the case.

> Maybe there are two concepts here?

Given that we can't agree, that seems very likely :grin: 

Therefore my proposal is that we keep ResourceClass and support an optional reference to it in the claim and one optional reference in the request. That should be expressive enough explore what works and what doesn't. Because it's a one-of, we can add back multi-inheritance or other types in the future.

I'm undecided about default requests in the class. That is needed for "highly expressive classes" (cannot do a class which merely needs to be referenced to get two different devices without them).

Reply:
@johnbelamaric at least in classic DRA I didn't actually have much of a use-case for classes.

Instead of defining a set of classes, I relied on the fact that I could define a set of custom CRDs to be used as `ClaimParameters` for a claim. The claim itself was merely a vessel to provide a reference to (one of) my vendor-specific CRDs. Using different CRD variants, I could control the exact semantics around what devices a user could request with each CRD, as well as the exact set of knobs for selecting / configuring devices of that type.

Two representative examples include:
(1) it is unsupported to select more than 1 MIG device, so I left the `count` field out of the `MigDeviceClaimParameters` CRD, and
(2) it is unsupported to time-slice or run MPS for a GPU managed by a vfio driver, so I left these `sharing` options out of the `VfioGpuClaimParameters` object.

Reply:
Here's another, simpler alternative that is more in line with John's proposal:
- We use DeviceClass instead of ResourceClass.
- DeviceClass only has fields related to requests:
  - per-device configuration
  - device requirements
- DeviceClasses can be referenced in a claim request, but not in the claim itself.
- Using a class remains optional.

Much easier to understand and explain. It's also more limited. There's no way to pre-define constraints between multiple devices. Users have to put those into their claims. I'm fine with this.

Reply:
Because classes are optional, we need a different mechanism to limit which devices are accessible within which namespace. See https://github.com/kubernetes-sigs/wg-device-management/pull/14/files#r1613095446 for a proposal.

Reply:
Even if classes were mandatory, we then still would need an access control mechanism. I prefer to decouple "device access control" from "privileged device configuration access control" (aka access to classes). They seem orthogonal to me.

Reply:
> a) we need a way to do quota

Agreed. We also need to do it at allocation time, to accurately reflect what the user actually is getting.

If we were to base quota on classes, the classes would have to be accurate: if class is "a 10GiB GPU or larger", then users can game the system by using that class and hoping that they get something bigger. Therefore I believe that quota should be based on device attributes, not classes.

> b) I'm not sure classes are really valuable

The current proposal (DeviceClass) seems useful to me, but I am also okay with dropping them as long as we design the API so that we can add them back safely through a separate KEP.

We could "reserve" fields for future use:
```
type ResourceClaimSpec struct {

     // At the moment, ClaimClassRef is always. But clients *must* check
     // that it really is nil and error out with "unsupported class reference" if it is non-nil
     // because a future extension might populate this field.
    *ClaimClassRef `json:",inline"`
}

// A one-of, always empty at the moment and thus not usable yet.
type ClaimClassRef struct {
    // FUTURE EXTENSION(s):
    // ClaimClassName *string `json:"claimClassName"`
}
```

Same for a request instead of defining a `deviceClassName`. Note that although the struct is now called `ClaimClassRef`, the actual field name is only going to be defined in the future and even the struct can be renamed. Nothing of this will appear in the current user API, it's just Go code.

We should only do this judiciously. It puts a burden on clients which might never pay off. But it is a way for us to punt on implementing something now that we want (or might want) to add in the future.

The error message in current clients needs to know what the field is for. We could even avoid that with:
```
type ResourceClaimSpec struct {

    // Clients must check for nil. If non-nil, they must error out with
    // "unsupported field: <FooField.FieldName>.
    *FooField `json:",inline"`
}

// Type must be set plus exactly one other field.
type ClaimClassRef struct {
    FieldName string

    // FUTURE EXTENSION(s):
    // ClaimClassName *string `json:"claimClassName"`
}
```

But that would make the future API weird: even if we set the FieldName in the API server as a default, it would be user-visible.